### PR TITLE
Add ability set specific language when render file

### PR DESCRIPTION
**Pull request by thenotsoft**

### CODE REVIEW

**Review by devanych — APPROVED** [review]

File: src/ViewTrait.php [hunk 1]
```diff
@@ -470,7 +501,7 @@ public function renderFile(string $viewFile, array $parameters = []): string
      */
     public function localize(string $file, ?string $language = null, ?string $sourceLanguage = null): string
     {
-        $language = $language ?? $this->state->getLanguage();
+        $language = $language ?? $this->language ?? $this->state->getLanguage();
```
Comment (vjik) [inline src/ViewTrait.php:504]: Too implicit behavior. Suggest remove `withLanguage()` and leave only new parameter "language" in methods.

Comment (rustamwin) [reply src/ViewTrait.php:504]: I agree with @vjik as this is only used for rendering which already has such a parameter.

File: src/ViewTrait.php [hunk 2]
```diff
@@ -426,7 +443,7 @@ public function renderFile(string $viewFile, array $parameters = []): string
         }
 
         if (is_file($viewFile)) {
-            $viewFile = $this->localize($viewFile);
+            $viewFile = $this->localize($viewFile, $this->localeState->getLanguage());
```
Comment (rustamwin) [inline src/ViewTrait.php:446]: Passing from locale state is redundant. Already used as a fallback language in this method.

File: src/WebView.php [hunk 3]
```diff
@@ -164,7 +167,7 @@ public function endPage(bool $ajaxMode = false): void
             sprintf(self::PLACEHOLDER_BODY_END, $this->getPlaceholderSignature()) => $this->renderBodyEndHtml($ajaxMode),
         ]);
 
-        $this->clear();
+        $this->state->clear();
```
Comment (vjik) [inline src/WebView.php:170]: Suggest leave `$this->clear()` as is, and then discuss about this in another issue/PR.

File: src/ViewTrait.php [hunk 4]
```diff
@@ -514,6 +531,7 @@ public function clear(): void
     {
         $this->viewFiles = [];
         $this->state->clear();
+        $this->localeState = new LocaleState();
```
Comment (vjik) [inline src/ViewTrait.php:534]: May be add method `LocaleState::clear()` and use him here?
Only suggestion for consistent with `ViewState` and `WebViewState`.

File: src/View.php [hunk 5]
```diff
@@ -27,6 +28,7 @@ final class View implements ViewInterface
     use ViewTrait;
 
     private ViewState $state;
+    private LocaleState $localeState;
```
Comment (vjik) [inline src/View.php:31]: In `View::withClearedState()` method need clear locale state.

File: src/WebView.php [hunk 6]
```diff
@@ -40,6 +41,7 @@ final class WebView implements ViewInterface
     use ViewTrait;
 
     private WebViewState $state;
+    private LocaleState $localeState;
```
Comment (vjik) [inline src/WebView.php:44]: In `WebView::withClearedState()` method need clear locale state.

File: src/State/StateTrait.php [hunk 7]
```diff
@@ -17,7 +17,6 @@
 trait StateTrait
 {
     private ?Theme $theme = null;
-    private string $language = 'en';
```
Comment (vjik) [inline src/State/StateTrait.php:20]: Need fix docs in readme and describe changes.

File: src/State/LocaleState.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\View\State;
+
+final class LocaleState
```
Comment (vjik) [inline src/State/LocaleState.php:7]:
```suggestion
/**
 * @internal
 */
final class LocaleState
```

File: tests/ViewTest.php [hunk 9]
```diff
@@ -233,6 +233,57 @@ public function testLocalize(): void
...
+        FileHelper::ensureDirectory($this->tempDirectory);
+        FileHelper::ensureDirectory("$this->tempDirectory/es");
+        file_put_contents("$this->tempDirectory/file.php", 'test en render');
+        file_put_contents("$this->tempDirectory/es/file.php", 'test es render');
+
+        $view = $this->createViewWithBasePath($this->tempDirectory);
+
+        $this->assertSameIgnoringSlash(
+            "test es render",
+            $view->withLanguage('es')->render('/file'),
```
Comment (vjik) [inline tests/ViewTest.php:263]: Need test to immutable also.

**Comment by vjik** [comment]
Need line in change log also

File: src/ViewTrait.php (hunk 4 above)
Comment (thenotsoft) [reply src/ViewTrait.php:534]: This is only for internal use, I don’t see much point in creating an additional method when you can create a new instance.

**Comment by samdark** [comment]
Would it be properly passed to sub-views?

**Review by samdark — APPROVED** [review]

**Review by rustamwin — APPROVED** [review]

File: src/WebView.php (hunk 3 above)
Comment (samdark) [reply src/WebView.php:171]: Discussed it internally. It's OK as in the PR.

File: README.md [hunk 10]
```diff
@@ -90,6 +88,44 @@ final class LastPosts extends Widget
 }
 ```
 
+### LocaleState of `View` and `WebView` services
+`LocaleState` like the previous state is not cloned when services are cloned. You can change the local by using `setLanguage()`, which will be applied to all other instances and to the old ones. If you need to change the locale only for a new instance without changing the locale globally, then you can use the immutable method `withLanguage()`, changing the locale also be applied to all sub-renders.
```
Comment (samdark) [inline README.md:92]:
```suggestion
`LocaleState` like the previous state is not cloned when services are cloned. You can change the locale by using `setLanguage()`, which will be applied to all other instances including existing ones. If you need to change the locale only for a single instance, you can use the immutable `withLanguage()` method. Locale will be applied to all views rendered within views with `render()` calls.
```

File: README.md [hunk 11]
```diff
@@ -46,14 +46,12 @@ data.
 `View` service:
 - parameters,
 - blocks,
-- theme,
-- language.
```
Comment (vjik) [inline README.md:50]: Why removed? Language in state also.

File: README.md [hunk 12]
```diff
@@ -46,14 +46,12 @@ data.
...
+- theme.
 
 `WebView` service:
 - parameters,
 - blocks,
 - theme,
-- language,
```
Comment (vjik) [inline README.md:56]: Why removed? Language in state also.

File: README.md [hunk 13]
```diff
@@ -90,6 +88,44 @@ final class LastPosts extends Widget
 }
 ```
 
+### LocaleState of `View` and `WebView` services
```
Comment (vjik) [inline README.md:91]:
```suggestion
#### Language in state
```

This is subheading of "State of `View` and `WebView` services" part.

File: README.md [hunk 14]
```diff
@@ -90,6 +88,44 @@ final class LastPosts extends Widget
...
+`LocaleState` like the previous state is not cloned when services are cloned. You can change the locale by using `setLanguage()`, which will be applied to all other instances including existing ones. If you need to change the locale only for a single instance, you can use the immutable `withLanguage()` method. Locale will be applied to all views rendered within views with `render()` calls.
+
+example with mutable method:
```
Comment (vjik) [inline README.md:94]:
```suggestion
Example with mutable method:

```

File: README.md [hunk 15]
```diff
@@ -90,6 +88,44 @@ final class LastPosts extends Widget
...
+    }
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        ...
+        $this->view->setLanguage($locale);
+        ...
+    }
+}
+````
+or immutable method:
```
Comment (vjik) [inline README.md:116]:
```suggestion

Example with immutable method:

```

File: README.md [hunk 16]
```diff
@@ -90,6 +88,44 @@ final class LastPosts extends Widget
...
+    public function __construct (WebView $view) {
+        $this->view = $view;
+    }
+    public function index() {
+        return $this->view->withLanguage('es')->render('index');
+    }
+}
+```
+
 To get a deep cloned `View` or `WebView` use `withClearedState()`:
```
Comment (vjik) [inline README.md:129]: Deep clone should be before locale state part.

File: README.md [hunk 17]
```diff
@@ -90,6 +88,44 @@ final class LastPosts extends Widget
 }
 ```
 
+### LocaleState of `View` and `WebView` services
+`LocaleState` like the previous state is not cloned when services are cloned. You can change the locale by using `setLocale()`, which will be applied to all other instances including existing ones. If you need to change the locale only for a single instance, you can use the immutable `withLocale()` method. Locale will be applied to all views rendered within views with `render()` calls.
```
Comment (vjik) [inline README.md:92]:
```suggestion

You can change the locale by using `setLocale()`, which will be applied to all other instances that used current state 
including existing ones. If you need to change the locale only for a single instance, you can use the immutable 
`withLocale()` method. Locale will be applied to all views rendered within views with `render()` calls.
```

`LocaleState` is internal class. We shouldn't mention him.

File: src/ViewInterface.php [hunk 18]
```diff
@@ -43,13 +43,13 @@ public function withBasePath(string $basePath): self;
...
-     * Returns a new instance with the specified source language.
+     * Returns a new instance with the specified source locale.
      *
-     * @param string $language The source language.
+     * @param string $locale The source locale.
      *
      * @return static
      */
-    public function withSourceLanguage(string $language): self;
+    public function withSourceLocale(string $locale): self;
```
Comment (vjik) [inline src/ViewInterface.php:52]: This should be mention in changelog.

File: src/ViewInterface.php [hunk 19]
```diff
@@ -96,13 +96,22 @@ public function withPlaceholderSalt(string $salt): self;
...
-     * Set the specified language code.
+     * Set the specified locale code.
      *
-     * @param string $language The language code.
+     * @param string $locale The locale code.
      *
      * @return static
      */
-    public function setLanguage(string $language): self;
+    public function setLocale(string $locale): self;
```
Comment (vjik) [inline src/ViewInterface.php:105]: This should be mention in changelog.

File: src/ViewInterface.php [hunk 20]
```diff
@@ -96,13 +96,22 @@ public function withPlaceholderSalt(string $salt): self;
...
+
+    /**
+     * Set the specified locale code.
+     *
+     * @param string $locale The locale code.
+     *
+     * @return static
+     */
+    public function withLocale(string $locale): self;
```
Comment (vjik) [inline src/ViewInterface.php:114]: This should be renamed in changelog.

File: README.md (hunk 13 above)
Comment (samdark) [reply README.md:91]: It's locale, not language.

File: README.md (hunk 16 above)
Comment (thenotsoft) [reply README.md:134]: Why?, deep clone also resets locale state.

File: CHANGELOG.md [hunk 21]
```diff
@@ -2,6 +2,9 @@
 
 ## 5.0.2 under development
 
+- New #199: Add immutable method `ViewInterface::withLocale()` that set locale (@thenotsoft, @vjik, @samdark)
+- Chg #199: Renamed method `ViewInterface::setLanguage()` to `ViewInterface::setLocale` (@thenotsoft, @samdark)
```
Comment (vjik) [inline CHANGELOG.md:6]:
```suggestion
- Chg #199: Renamed method `ViewInterface::setLanguage()` to `ViewInterface::setLocale()` (@thenotsoft, @samdark)
```

File: CHANGELOG.md [hunk 22]
```diff
@@ -2,6 +2,9 @@
...
+- Chg #199: Renamed method `ViewInterface::withSourceLanguage()` to `ViewInterface::withSourceLocale` (@thenotsoft, @samdark)
```
Comment (vjik) [inline CHANGELOG.md:7]:
```suggestion
- Chg #199: Renamed method `ViewInterface::withSourceLanguage()` to `ViewInterface::withSourceLocale()` (@thenotsoft, @samdark)
```

File: README.md [hunk 23]
```diff
@@ -46,14 +46,14 @@ data.
 `View` service:
 - parameters,
 - blocks,
-- theme,
-- language.
+- theme
```
Comment (vjik) [inline README.md:49]:
```suggestion
- theme,
```

File: README.md [hunk 24]
```diff
@@ -90,6 +90,49 @@ final class LastPosts extends Widget
 }
 ```
 
+#### Locale state
+You can change the locale by using `setLocale()`, which will be applied to all other instances that used current state
```
Comment (vjik) [inline README.md:94]:
```suggestion

You can change the locale by using `setLocale()`, which will be applied to all other instances that used current state
```

File: README.md (hunk 16 above)
Comment (vjik) [reply README.md:136]: Then need heading. Now seems that deep clone into "Locale state".

File: README.md [hunk 25]
```diff
@@ -90,6 +90,51 @@ final class LastPosts extends Widget
...
+    public function __construct (WebView $view) {
+        $this->view = $view;
+    }
+    public function index() {
+        return $this->view->withLocale('es')->render('index');
+    }
+}
+```
+
+#### Reset state
```
Comment (vjik) [inline README.md:137]:
```suggestion
#### Reset state

```

**Pull request merged**